ttljob: add metrics to row level TTL

otan opened the pull request:
This commit adds metrics about the row-level TTL job. Note new metrics
are created on a per-TTL table basis, which forms the basis of the
"relation" label. This allows monitoring of TTL jobs per table.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76701)
<!-- Reviewable:end -->

miretskiy commented on pkg/sql/ttl/ttljob/ttljob.go:
```diff
@@ -63,12 +66,104 @@ type rowLevelTTLResumer struct {
+			},
+		),
+	}
+	addMetricStructToRegistry(ret)
```
So, previously, when I attempted to do this kind of dynamic registration, scoped by name (relationship) in this case, I got some well reasoned pushback from @andreimatei .  I wonder if he has thoughts on this.
Also, I don't know that registry supports adding metrics struct containing maps -- I see it supports array of metric structs, but definitely not maps.

I suspect that perhaps the thing to do here might be to use Agg* metrics, like we do in changefeedccl/metrics?

otan replied:
I don't see a way of doing this using aggmetric as we don't know the table names upfront. Also note we only keep using the map to reference the metrics - the MakeRowLevelTTLMetrics function adds it to the server registry. The map is simply used to reference the metric struct internally.

otan replied:
ah i misread aggmetric - i'll look later

otan replied:
The aggmetric labels don't seem to show up in /_status/vars - - do you know where they actually show up?

miretskiy commented on pkg/jobs/registry.go:
```diff
@@ -186,6 +186,7 @@ func MakeRegistry(
 	sqlInstance sqlliveness.Instance,
 	settings *cluster.Settings,
 	histogramWindowInterval time.Duration,
+	addMetricStruct func(interface{}),
```
I don't think you need this argument anymore, don't you?
It's kinda weird by job registry is itself a MetricStruct -- it returns it's Metrics{} -- and since you just added another metric struct there -- all the initialization should happen auto-magically.

rafiss commented on pkg/sql/ttl/ttljob/ttljob.go:
```diff
@@ -63,12 +68,127 @@ type rowLevelTTLResumer struct {
+
+	mu struct {
+		syncutil.Mutex
+		m map[string]rowLevelTTLMetrics // map[int]rowLevelTTLMetrics
```
seems like an old comment?

rafiss commented on pkg/sql/ttl/ttljob/ttljob.go:
```diff
@@ -63,12 +68,127 @@ type rowLevelTTLResumer struct {
+			metric.Metadata{
+				Name:        "jobs.row_level_ttl.range_total_duration",
+				Measurement: "seconds",
+				Unit:        metric.Unit_NANOSECONDS,
```
nit: measurement says seconds but unit says nanoseconds

rafiss commented on pkg/sql/ttl/ttljob/ttljob.go:
```diff
@@ -127,17 +248,49 @@ func (t rowLevelTTLResumer) Resume(ctx context.Context, execCtx interface{}) err
 		return err
 	}
 
+	metrics := p.ExecCfg().JobRegistry.MetricsStruct().RowLevelTTL.(*RowLevelTTLAggMetrics).loadMetrics(name)
```
do we have the pattern of keying metrics by table name elsewhere? what if the table name changes?

otan replied:
See below

rafiss commented on pkg/sql/ttl/ttljob/ttljob.go:
```diff
@@ -127,17 +248,49 @@ func (t rowLevelTTLResumer) Resume(ctx context.Context, execCtx interface{}) err
 		return err
 	}
 
+	metrics := p.ExecCfg().JobRegistry.MetricsStruct().RowLevelTTL.(*RowLevelTTLAggMetrics).loadMetrics(name)
```
do we have the pattern of keying metrics by table name elsewhere? what if the table name changes?

actually i may not understand what `loadMetrics` does

otan replied:
yeah it converts an aggmetric with no label attached to a simple metric which does.

otan replied:
See other usages of aggmetric, from memory slimetrics was an example.

rafiss commented on pkg/sql/ttl/ttljob/ttljob.go:
```diff
@@ -127,17 +248,49 @@ func (t rowLevelTTLResumer) Resume(ctx context.Context, execCtx interface{}) err
+		if err != nil {
+			return err
+		}
+		schemaDesc, err := descs.GetImmutableSchemaByID(
```
agh i wish it were easier to make a table name without having to fetch these descriptors :/

otan replied:
Me too...

rafiss commented on pkg/sql/ttl/ttljob/ttljob.go:
```diff
@@ -127,17 +248,49 @@ func (t rowLevelTTLResumer) Resume(ctx context.Context, execCtx interface{}) err
+		if err != nil {
+			return err
+		}
+		schemaDesc, err := descs.GetImmutableSchemaByID(
```
agh i wish it were easier to make a table name without having to fetch these descriptors :/

rafiss reviewed (changes requested):
minor nits and questions

rafiss commented:
double post'd!

otan committed:
ttljob: add metrics to row level TTL

This commit adds metrics about the row-level TTL job. Note new metrics
are created on a per-TTL table basis, which forms the basis of the
"relation" label. This allows monitoring of TTL jobs per table.

Release note: None

otan commented:
bors r+

assuming nits are non-blocking, they've all been addressed!

miretskiy commented on pkg/sql/ttl/ttljob/ttljob.go:
```diff
@@ -72,12 +77,133 @@ type rowLevelTTLResumer struct {
+}
+
+func makeRowLevelTTLAggMetrics(histogramWindowInterval time.Duration) metric.Struct {
+	sigFigs := 2
```
btw, are we sure we want sigFigs=2?   Wouldn't 1 be sufficient?
